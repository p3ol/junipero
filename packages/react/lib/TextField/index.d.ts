import React, { ComponentPropsWithRef, MutableRefObject } from "react";

export declare type TextFieldRef = {
  innerRef: MutableRefObject<any>;
  inputRef: MutableRefObject<any>;
  value: String;
  valid: Boolean;
  dirty: Boolean;
  focused: Boolean;
  focus: () => void;
  blur: () => void;
  reset: () => void;
  setDirty: (dirty: Boolean) => void;
  isJunipero: Boolean;
};

declare interface TextFieldProps extends ComponentPropsWithRef<any> {
  autoFocus?: Boolean;
  children: React.ReactNode;
  className: String;
  valid: Boolean;
  value?: String;
  disabled: Boolean;
  required?: Boolean;
  type?: String;
  onBlur?: (event: Event) => void;
  onFocus?: (event: Event) => void;
  onChange?: (props: { value: String; valid: Boolean; dirty: Boolean }) => void;
  onValidate?: (
    val: String,
    options: { required?: Boolean; dirty?: Boolean }
  ) => Boolean;
  ref?: MutableRefObject<TextFieldRef | undefined>;
}

declare function TextField(props: TextFieldProps): JSX.Element;

export default TextField;
