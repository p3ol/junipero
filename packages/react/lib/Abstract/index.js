import { classNames } from '@junipero/core';

import { useFieldControl } from '../hooks';

const Abstract = ({ className, ...rest }) => {
  const { valid = true, dirty = false, focused = false } = useFieldControl();

  return (
    <div
      { ...rest }
      className={classNames(
        'junipero abstract',
        { invalid: !valid && !focused && dirty },
        className
      )}
    />
  );
};

Abstract.displayName = 'Abstract';

export default Abstract;
