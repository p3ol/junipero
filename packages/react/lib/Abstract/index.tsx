import { classNames } from '@junipero/core';
import { ComponentPropsWithoutRef, ReactNode } from 'react';

import { useFieldControl } from '../hooks';

declare interface AbstractProps extends ComponentPropsWithoutRef<any> {
  className?: string;
  children?: ReactNode | JSX.Element;
}

const Abstract = (
  {
    className,
    ...rest
  }: AbstractProps) => {
  const { valid = true, dirty = false, focused = false } = useFieldControl();

  return (
    <div
      { ...rest }
      className={classNames(
        'junipero abstract',
        { invalid: !valid && !focused && dirty },
        className
      )}
    />
  );
};

Abstract.displayName = 'Abstract';

export default Abstract;
