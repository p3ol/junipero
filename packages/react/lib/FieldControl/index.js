import { useCallback, useReducer } from 'react';
import { mockState } from '@junipero/core';

import { FieldControlContext } from '../contexts';

const FieldControl = props => {
  const [state, dispatch] = useReducer(mockState, {
    valid: true,
    dirty: false,
    focused: false,
  });

  const getContext = useCallback(() => ({
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    update: dispatch,
  }), [state.valid, state.dirty, state.focused]);

  return (
    <FieldControlContext.Provider value={getContext()} {...props} />
  );
};

FieldControl.displayName = 'FieldControl';

export default FieldControl;
