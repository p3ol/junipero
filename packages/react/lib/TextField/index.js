import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
