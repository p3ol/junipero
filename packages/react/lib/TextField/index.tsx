import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
