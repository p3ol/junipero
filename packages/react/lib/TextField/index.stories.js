import { action } from '@storybook/addon-actions';

import Label from '../Label';
import Abstract from '../Abstract';
import TextField from './index';
import FieldControl from '../FieldControl';

export default { title: 'react/TextField' };

export const basic = () => (
  <TextField placeholder="First name" onChange={action('change')} />
);

export const basicMultiline = () => (
  <TextField
    rows={10}
    label="First name"
    placeholder="First name"
    onChange={action('change')}
  />
);

export const withLabelAndAbstract = () => (
  <FieldControl>
    <Label className="info" htmlFor="firstname">First name</Label>
    <TextField
      id="firstname"
      placeholder="Type your first name"
      onChange={action('change')}
      onValidate={() => false}
    />
    <Abstract className="info">
      Here is a little help writing your name
    </Abstract>
  </FieldControl>
);

export const autoFocused = () => (
  <TextField autoFocus onChange={action('change')} />
);

export const disabled = () => (
  <TextField disabled value="test" />
);

export const withValidation = () => (
  <TextField
    label="Age"
    placeholder="Age"
    onValidate={val => /^[0-9]+$/g.test(val)}
  />
);
