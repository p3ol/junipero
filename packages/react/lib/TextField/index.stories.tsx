import { action } from 'storybook/actions';

import Label from '../Label';
import Abstract from '../Abstract';
import FieldControl from '../FieldControl';
import TextField from './index';

export default { title: 'react/TextField' };

export const basic = () => (
  <TextField placeholder="First name" onChange={action('change')} />
);

export const basicMultiline = () => (
  <TextField
    rows={10}
    label="First name"
    placeholder="First name"
    onChange={action('change')}
  />
);

export const withLabelAndAbstract = () => (
  <FieldControl>
    <Label className="info" htmlFor="firstname">First name</Label>
    <TextField
      id="firstname"
      placeholder="Type your first name"
      onChange={action('change')}
      onValidate={() => false}
    />
    <Abstract className="info">
      Here is a little help writing your name
    </Abstract>
  </FieldControl>
);

export const autoFocused = () => (
  <TextField autoFocus onChange={action('change')} />
);

export const disabled = () => (
  <TextField disabled value="test" />
);

export const withValidation = () => (
  <TextField
    label="Age"
    placeholder="Age"
    onValidate={(val: string) => /^[0-9]+$/g.test(val)}
  />
);

export const typeNumber = () => (
  <TextField
    className="field noscroll"
    label="Age"
    placeholder="Age"
    type="number"
    min={1}
    value={14}
    // max={100}
    onChange={action('change')}
  />
);

export const withOnWheelOverrided = () => (
  <TextField
    className="field noscroll"
    label="Age"
    placeholder="Age"
    type="number"
    onWheel={action('wheel')}
  />
);
