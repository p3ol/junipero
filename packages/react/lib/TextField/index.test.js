import { createRef } from 'react';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { blur, reset } from '~test-utils';
import Label from '../Label';
import Abstract from '../Abstract';
import FieldControl from '../FieldControl';
import TextField from './';

describe('<TextField />', () => {
  it('should render', () => {
    const { unmount, container } = render(<TextField placeholder="Name" />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should be invalid if validation fails', async () => {
    const user = userEvent.setup();
    const { unmount, container } = render(
      <TextField
        placeholder="Name"
        onValidate={() => false}
      />
    );

    const input = container.querySelector('input');
    await user.type(input, 'John');
    await blur(input);
    expect(container).toMatchSnapshot();

    unmount();
  });
  it('should return number if type is number', async () => {
    const onChangeMock = jest.fn();
    const user = userEvent.setup();
    const { unmount, container } = render(
      <TextField
        type="number"
        min={1}
        onChange={onChangeMock}
      />
    );

    const input = container.querySelector('input');
    await user.type(input, '3');
    await blur(input);
    
    expect(onChangeMock).toHaveBeenCalledWith(
      expect.objectContaining({ value: 3 })
    );

    unmount();
  });
  
  it('should be invalid if number is out of range', async () => {
    const user = userEvent.setup();
    const { unmount, container } = render(
      <TextField
        type="number"
        min={1}
      />
    );

    const input = container.querySelector('input');
    await user.type(input, '-1');
    await blur(input);

    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should allow to be used with a FieldControl', async () => {
    const user = userEvent.setup();
    const { unmount, container } = render(
      <FieldControl>
        <Label htmlFor="name">Name</Label>
        <TextField id="name" placeholder="Name" onValidate={() => false} />
        <Abstract>Enter your name</Abstract>
      </FieldControl>
    );

    const input = container.querySelector('input');
    await user.type(input, 'John');
    await blur(input);

    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should allow to reset the field', async () => {
    const user = userEvent.setup();
    const ref = createRef();
    const onChange = jest.fn();
    const { unmount, container } = render(
      <TextField ref={ref} value="John" onChange={onChange} />
    );
    const input = container.querySelector('input');

    await user.clear(input);
    await user.type(input, 'Jane');
    await blur(input);

    expect(container).toMatchSnapshot();
    expect(onChange)
      .toHaveBeenLastCalledWith(expect.objectContaining({ value: 'Jane' }));

    await reset(ref.current);
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should not allow to change value when disabled', async () => {
    const user = userEvent.setup();
    const onChange = jest.fn();
    const { unmount, container } = render(
      <TextField onChange={onChange} disabled />
    );
    const input = container.querySelector('input');
    await user.type(input, 'Jane');
    await blur(input);

    expect(onChange).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should render a textarea if rows is set to > 1', () => {
    const { unmount, container } = render(<TextField rows={2} />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should blur on scroll', async () => {
    const user = userEvent.setup();
    const { unmount, container } = render(<TextField type="number" />);

    const input = container.querySelector('input');
    await user.click(input);
    expect(container).toMatchSnapshot('focused');
    fireEvent.wheel(container.querySelector('.junipero.text-field input'), {
      deltaY: 30,
    });
    expect(container).toMatchSnapshot('unfocused');
    unmount();
  });

  it('should do the given action on scroll if onWheel is set', async () => {
    const user = userEvent.setup();
    const customWheel = jest.fn();
    const { unmount, container } = render(
      <TextField type="number" onWheel={customWheel} />
    );

    const input = container.querySelector('input');
    await user.click(input);
    expect(container).toMatchSnapshot('focused');
    fireEvent.wheel(container.querySelector('.junipero.text-field input'), {
      deltaY: 30,
    });
    expect(customWheel).toHaveBeenCalled();
    expect(container).toMatchSnapshot('always focused');
    unmount();
  });
});
