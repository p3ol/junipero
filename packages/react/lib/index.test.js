import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
