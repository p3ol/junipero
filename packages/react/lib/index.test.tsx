import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
