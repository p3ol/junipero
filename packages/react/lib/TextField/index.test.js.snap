// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should allow to be used with a FieldControl 1`] = `
<div>
  <label
    class="junipero label invalid"
    for="name"
  >
    Name
  </label>
  <div
    class="junipero text-field dirty invalid"
  >
    <input
      class="field"
      id="name"
      placeholder="Name"
      type="text"
      value="John"
    />
  </div>
  <div
    class="junipero abstract invalid"
  >
    Enter your name
  </div>
</div>
`;

exports[`<TextField /> should allow to reset the field 1`] = `
<div>
  <div
    class="junipero text-field dirty valid"
  >
    <input
      class="field"
      type="text"
      value="Jane"
    />
  </div>
</div>
`;

exports[`<TextField /> should allow to reset the field 2`] = `
<div>
  <div
    class="junipero text-field pristine valid"
  >
    <input
      class="field"
      type="text"
      value="John"
    />
  </div>
</div>
`;

exports[`<TextField /> should be invalid if validation fails 1`] = `
<div>
  <div
    class="junipero text-field dirty invalid"
  >
    <input
      class="field"
      placeholder="Name"
      type="text"
      value="John"
    />
  </div>
</div>
`;

exports[`<TextField /> should not allow to change value when disabled 1`] = `
<div>
  <div
    class="junipero text-field pristine valid empty disabled"
  >
    <input
      class="field"
      disabled=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`<TextField /> should render 1`] = `
<div>
  <div
    class="junipero text-field pristine valid empty"
  >
    <input
      class="field"
      placeholder="Name"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`<TextField /> should render a textarea if rows is set to > 1 1`] = `
<div>
  <div
    class="junipero text-field pristine valid empty"
  >
    <textarea
      class="field"
      rows="2"
    />
  </div>
</div>
`;
